source/common/http/http2/codec_impl.cc

```diff
@@ -349,6 +354,39 @@ void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadat
   }
 }
 
+void ConnectionImpl::StreamImpl::processBufferedData() {
+  ENVOY_CONN_LOG(debug, "Stream {} processing buffered data.", parent_.connection_, stream_id_);
+
+  // We should stop processing buffered data if either
+  // 1) Buffers become overrun
+  // 2) The stream ends up getting reset
+  // Both of these can end up changing as a result of processing buffered data.
+  bool continue_processing = !buffersOverrun() && !reset_reason_.has_value();
+
+  if (stream_manager_.body_buffered_ && continue_processing) {
+    if (stream_manager_.data_end_stream_) {
+      remote_end_stream_ = true; // restore the deferred end_stream from the remote.
```

Comment:
I'd be inclined to keep remote_end_stream as it was, but change the call sites which use it to use a utility function sendEndStream() which does the needful.  Alternately you could name remote_end_stream to send_remote_end_stream but having received a remote end stream and having this false is likely to cause debug confusion.

Reply:
great idea, done :)